Add configuration and implement part of the client

adrianfusco opened the pull request:
- `go.mod` & `go.sum`: Used [go-lumber](https://github.com/elastic/go-lumber) done by elastic that allows connecting to the logstash server and send the messages using the Lumberjack protocol over TCP. At this way we can communicate with logstash using the `beats` input plugin. Also added other dependencies of this one.
- `configs/config.go`: Added simple package to struct the information related to the configuration, parse the YAML config file and operate over the duration of the recheck.
- `gz-beat-shipper-conf.yml`: name of the configuration file used for the configuration directives.
- `pkg/filehandler/filehandler.go`: Added method to get the content of a GNU Zip file.
- `cmd/shipper/main.go`: Add all the implementations done in other PRs and this one. 

With this PR it starts to work:

- Let's say we have the following GZ file:

```
[2022-10-10 22:38:50] {afuscoar@afuscoar} (~)$ -> zcat /rhos-infra-dev-netapp/jenkins-logs/rcj/test/sub-test/sub-sub-test/var/log/extra/rpm-list.txt.gz 
package1.rpm
package2.rpm
package3.rpm
```

- We run it:

```
[2022-10-11 00:00:30] {afuscoar@afuscoar} (~/.Redhat/Projects/github/gz-beat-shipper) (feature/shipper-messages)$ -> go run cmd/shipper/main.go 

# Start processing the file:
2022/10/11 00:13:54 main.go:93: Processing: /rhos-infra-dev-netapp/jenkins-logs/rcj/test/sub-test/sub-sub-test/var/log/extra/rpm-list.txt.gz

# Add the file to the local registry:
2022/10/11 00:13:54 registry.go:59: Added to registry: {/rhos-infra-dev-netapp/jenkins-logs/rcj/test/sub-test/sub-sub-test/var/log/extra/rpm-list.txt.gz}

# There aren't new entries so nothing else to process:
2022/10/11 00:13:59 main.go:60: There aren't gz files to process for /rhos-infra-dev-netapp/jenkins-logs/rcj/*/*/*/var/log/extra/rpm-list.txt.gz
2022/10/11 00:14:04 main.go:60: There aren't gz files to process for /rhos-infra-dev-netapp/jenkins-logs/rcj/*/*/*/var/log/extra/rpm-list.txt.gz
```
- Check the registry:

```
[2022-10-11 00:00:30] {afuscoar@afuscoar} (~/.Redhat/Projects/github/gz-beat-shipper) (feature/shipper-messages)$ -> cat shipper_registry.json | jq
{
  "ParsedFile": [
    {
      "name": "/rhos-infra-dev-netapp/jenkins-logs/rcj/test/sub-test/sub-sub-test/var/log/extra/rpm-list.txt.gz"
    }
  ]
}
```
- Logstash logs:

```
logstash         |                  "@version" => "1"
logstash         |                      "type" => "filebeat",
logstash         |                "@timestamp" => 2022-10-10T22:13:54.158Z,
logstash         |                   "message" => "package1.rpm\npackage2.rpm\npackage3.rpm\n",
logstash         |                  "@version" => "1"
```

adrianfusco commented:
I’ll also add in this PR:

- Close the connection. 
- Generates the binary
- Add the binary to systemD

cescgina commented on cmd/shipper/main.go:
```diff
@@ -1,16 +1,106 @@
 package main
 
 import (
+	"gz-beat-shipper/configs"
```
Question about Go, is there any convention regarding import orders? For example, first standard library, then other packages, then other parts of the own package?

adrianfusco replied:
I didn't find any convention for importing orders. I just see this commit as an example: https://github.com/golang/tools/commit/ed69e84b1518b5857a9f4e01d1f9cefdcc45246e 

Anyway, this is automatically ordered by the IDE using the linter that Go has per se.

jsanemet replied:
Convention is to order them in alphabetical order. Formatters will take care of that.

cescgina commented on cmd/shipper/main.go:
```diff
@@ -1,16 +1,106 @@
 func main() {
-	log.SetFlags(log.LstdFlags | log.Lshortfile)
+
+	config, r := setup()
```
the `r` variable could be named in a more descriptive manner

adrianfusco replied:
I've changed it. Thanks!

cescgina commented on configs/config.go:
```diff
@@ -0,0 +1,52 @@
+		log.Fatal(err)
+	}
+
+	s := t.Seconds()
```
What value would `t` have if an error is found?

adrianfusco replied:
If an error is found the code will exit in the `log.Fatal` part. If this code is executed successfully `time.ParseDuration(c.Recheck)` then this conversion can be done `t.Seconds()` and that's why it doesn't return error in the definition `func (d Duration) Seconds() float64 {`

cescgina replied:
Oh ok, I missed the fact that `log.Fatal` will cause the termination of the program. That seems counter-intuitive to me, but then the code is correct. Thanks!

cescgina commented on cmd/shipper/main.go:
```diff
@@ -1,16 +1,106 @@
+func getBatchToSend(gzFiles []string) []interface{} {
+	batch := make([]interface{}, 1)
+	i := 0
+	for _, gzFilePath := range gzFiles {
```
Shouldn't we keep the index from the for loop as `i`? Otherwise, the batch slice will keep only the last file found, won't it?

adrianfusco replied:
The index in 0 is just to create the batch slice since the first element. This `func` won't be ever called in case that there aren't files to process (checked before in `getGzNotProcessed`) and this is how `elastic` did it to create the event that can be received by logstash.

cescgina replied:
If we had  for example 3 files: file1, file2, file3 and call `getBatchToSend` with the 3, won't we end up with a batch that only contains file3? since all of them were assigned to `batch[0]`

adrianfusco replied:
Okay, I get what you say now, sorry for the misunderstood. I'm going to check this. I need to change the size of the slice that contain the event and also increment the iterator. Thanks for the explanation.

adrianfusco replied:
I've modified it using a slice without specifying the quantity of elements and just appending every new event inside.

Checked in logstash also that the batch of events is well received.

adrianfusco commented:
- Modified `.spec` file to create the dirs and move the binary, systemd service file and config file of the shipper.
- Modified workflow to generate both binary and rpm file and publish as artifacts.

Tested it:

```
# Downloaded, moved artifact and mount it in a docker:
[ROOT]afuscoar:/tmp/rpm$ -> docker run --network=elk_netElk -v /tmp/rpm/:/tmp/rpm -it rockylinux:9 bash

# Installed the package:
[root@f34289ab2baf rpm]# rpm -iv gz-beat-shipper-0.0.1-1.el9.x86_64.rpm 
Verifying packages...
Preparing packages...
gz-beat-shipper-0.0.1-1.el9.x86_64

# Changed the host to logstash - another docker running with logstash in the same network -
[root@f34289ab2baf rpm]# sed -i 's/localhost/logstash/g' /etc/gz-beat-shipper/gz-beat-shipper-conf.yml 

# Executed binary:
[root@f34289ab2baf rpm]# gz-beat-shipper 
2022/10/13 14:38:42 main.go:66: There aren't gz files to process for /rhos-infra-dev-netapp/jenkins-logs/rcj/*/*/*/var/log/extra/rpm-list.txt.gz

# Created GNU Zip file to test:
[root@5dafe2fdd57e /]# mkdir -p /rhos-infra-dev-netapp/jenkins-logs/rcj/test/subtest/subtest2/var/log/extra/
[root@5dafe2fdd57e rpm]# gzip /rhos-infra-dev-netapp/jenkins-logs/rcj/test/subtest/subtest2/var/log/extra/rpm-list.txt > /rhos-infra-dev-netapp/jenkins-logs/rcj/test/subtest/subtest2/var/log/extra/rpm-list.txt.gz

# Executed it:
[root@5dafe2fdd57e rpm]# gz-beat-shipper 
2022/10/13 14:47:10 main.go:107: Processing: /rhos-infra-dev-netapp/jenkins-logs/rcj/test/subtest/subtest2/var/log/extra/rpm-list.txt.gz
2022/10/13 14:47:10 registry.go:73: Added to registry: {/rhos-infra-dev-netapp/jenkins-logs/rcj/test/subtest/subtest2/var/log/extra/rpm-list.txt.gz}

# Received by logstash:

logstash         | {
logstash         |                "@timestamp" => 2022-10-13T14:47:10.088Z,
logstash         |                  "@version" => "1",
logstash         |                      "type" => "filebeat",
logstash         |                   "message" => "test.rpm\n",
logstash         |                      "tags" => [
logstash         |         [0] "beats_input_codec_plain_applied"
logstash         | }
```

IgnacioPro reviewed (approved):
+2

jsanemet commented on configs/config.go:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	s := t.Seconds()
+	return int32(s)
```
Seconds are usually stored in a floating point type, like float or double. By converting it into a integer you are losing a lot of precision: 1.4 -> 1.

If you'd really like to use an integer type, then time is usually measured in milliseconds instead.

jsanemet commented on configs/config.go:
```diff
@@ -0,0 +1,52 @@
+}
+
+func (c *Configuration) GetRecheckDuration() time.Duration {
+	recheckSeconds := c.GetRecheckInSeconds()
+	recheckDuration := time.Second * time.Duration(recheckSeconds)
+	return recheckDuration
```
No need to make amends, but you made things more complicated here than they could. On 'GetRecheckInSeconds', you get a 'time.Duration' which you then convert to seconds. Here, you take than conversion and transform it back into a duration.

You could just return 'time.ParseDuration(...)' here and have the other function call it instead.

jsanemet commented on configs/config.go:
```diff
@@ -0,0 +1,52 @@
+	Recheck  string `yaml:"recheck"`
+}
+
+func (c *Configuration) GetConfiguration() *Configuration {
```
You can leave this function as 'void', not much use for its returned value.

adrianfusco replied:
Sorry, can you explain me why the return type is `void`?

jsanemet replied:
You do nothing with the return value of this function. I am saying that you could return nothing altogether.

jsanemet commented on configs/config.go:
```diff
@@ -0,0 +1,52 @@
+	yaml "gopkg.in/yaml.v3"
+)
+
+const ConfigurationPath = "/etc/gz-beat-shipper/gz-beat-shipper-conf.yml"
```
My advise is to have this value be overridable somehow, with a CLI option or a configuration file, as this approach means that the user must have sudo rights in order to run the app.

JoseSilvaHermida reviewed (approved):
Very good job!

jsanemet reviewed (changes requested):
My major concern is that you are doing no error handling at all. You are crashing the app for all errors you see. I don't like that approach much, because it means that any piece of code on the application is able to shut it down. That will make it inflexible and hard to debug.

As an example, imagine that a caller wants to create a file on a certain directory. Under an error, the caller may want to try some other place or keep the data on memory. By crashing the app you give no option to react.
